cdb2api/cdb2api.c

```diff
@@ -3708,7 +3704,6 @@ static int cdb2_run_statement_typed_int(cdb2_hndl_tp *hndl, const char *sql,
 
     clear_responses(hndl);
 
-    hndl->sql = (char *)sql;
```

Comment:
So, prior to this change, we always kept track of the SQL (even when not in HA mode)?  It seems like that would be a  useful thing to preserve?  In that case, we could just always strdup() it?

Reply:
We always kept track of the last SQL even when not in HA mode, although only HA needs it to replay the query in case of a `cdb2_next_record()` failure. I agree that this would be useful to keep around, however calling `strdup()` on every single query sounds a bit too much overhead to me. By only strdup()'ing the SQL in HA mode, we could cut down a few CPU cycles for a non-HA query. A micro-optimization :)

Reply:
I don't think one call to strdup() per query (?) is going to be that costly.  It seems to me that *not* calling strdup() on an arbitrary string passed by the caller and then possibly trying to free() it could be problematic.  e.g. What if the caller passes in a string constant?  Conditionally calling strdup() could be OK *if* we keep track of whether the "sql" string needs to be freed using an additional flag of some kind.

Reply:
With this change, `hndl->sql` is either `NULL` (implied by `calloc()`-ing the handle), or a `strdup()`'d string. They both can be freed safely (although `free()`-ing NULL is a no op).